### PR TITLE
fix #86 adding verbose inheritance checks

### LINKED ISSUE #86: Make error messages more informative
>I found that when start and end dates are both in wrong class, the message will only mention the start - can we have error messages mention both are in wrong class?

_Originally posted by @Tina815 in https://github.com/Public-Health-Scotland/phsmethods/pull/85#discussion_r968554013_

In the functions `age_calculate` and `dob_from_chi` (and possibly some others) there is a check near the top which makes sure that the submitted parameters are the correct class/type.

Some parameters are intuitively paired (e.g. `min_date` and `max_date` in `dob_from_chi`). It would be good if the error message could check both of them and then report if either or both of them are the wrong class. Currently it checks one then if that is the wrong class it will error and report that, not checking the second one.

### PR DESCRIPTION
**Pull request by JoKPro**
Hi, 

I wrote a function that checks for the inheritance, prints out each argument that is lacking one of the specified classes and throws an exception. 
This is aiming to solve the issue #86 . 

I added a new file `make_inheritance_checks.R` that contains the function, together with some documentation. 
Hope this helps and let me know if there are any questions!

### CODE REVIEW

File: R/make_inheritance_checks.R [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  failures <- Filter(function(x) !is.null(x), failures)
+  if (!all(sapply(failures, is.null))) {
+    failures_out <- sapply(names(failures), FUN = function(element) {
+      if(!is.null(failures[[element]])) {
+        return(paste0(paste0(failures[[element]], collapse = "\n"), "\nClasses of ", "'", element, "': ", class(elements[[element]]), "\n"))
+      }
+    })
+    stop(cat(paste0("[make_inheritance_checks] failed in function call: ", "'", caller_func, "'", ":\n", paste0(failures_out, collapse = "\n"))))
+  }
+}
```
Comment (Moohan) [inline R/make_inheritance_checks.R:27]: Could these error messages be re-written using `cli` [](https://cli.r-lib.org/) please, which will keep them consistent with other error messages in the packages (see the 'dev' branch) and `cli` just looks great and has nice features for things like pluralisation, which is particularly relevant here.

File: R/make_inheritance_checks.R [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+#' Throws an exception if one element does not pass the checks.
+#' Element must have AT LEAST ONE of the specified classes to pass.
+#' @param elements a list object containing element_name=element pairs for each element. Element names must be provided or else they will be ignored.
+#' @param target_classes array of classes (string).
+
+make_inheritance_checks <- function(elements, target_classes) {
+  caller_func <- deparse(sys.calls()[[sys.nframe()-1]])
+  failures <- lapply(names(elements), FUN = function(element) {
+    if (is.null(elements[[element]])) {
+      return(paste0("Element ", "'", element, "'", " is null."))
```
Comment (Moohan) [inline R/make_inheritance_checks.R:11]:
```suggestion
      return(paste0("Argument", "'", element, "'", " is missing."))
```

**Review by Moohan — CHANGES_REQUESTED** [review]
Great work, and thanks for the PR 😊

It looks to me like it all works as expected, so my only comments are about improving the actual error message output.

**Comment by Moohan** [comment]
I've just seen the tests and lots are failing now, this seems to be because `dob_from_chi` has `NULL` by default for `min_date and `max_date` but your function is throwing an error when arguments are `NULL`.

File: R/make_inheritance_checks.R (hunk 1 above)
Comment (JoKPro) [reply R/make_inheritance_checks.R:27]: Thanks for the review! 
I will look into it.

**Comment by JoKPro** [comment]
I implemented the cli messages and added an argument `ignore_null` that lets you decide whether or not to ignore if checked arguments are NULL. Hope this resolves the issue.

File: R/make_inheritance_checks.R [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        return(failures[[argument]])
+      }
+    })
+    cli_abort(c(
+      "x" = "inheritance checks {col_red('failed')} in function call: {.strong {.code {caller_func}}} due to following errors:",
+      failures_out
+    ))
+  }
+  cli_alert_success("inheritance checks {col_green('passed')}")
+}
```
Comment (Moohan) [inline R/make_inheritance_checks.R:37]: You need to specify function from external packages as `pkg::fnc` (I think I've got them all but not sure)
```suggestion
make_inheritance_checks <- function(arguments, target_classes, ignore_null = T) {
  caller_func <- deparse(sys.calls()[[sys.nframe() - 1]])
  failures <- lapply(names(arguments), FUN = function(argument) {
    if (is.null(arguments[[argument]])) {
      if (ignore_null) {
        cli::cli_alert_info("{.fn make_inheritance_checks}: Argument {.arg { argument }} is {.val NULL}, but argument {.strong {.code ignore_null = TRUE}}.")
        return(NULL)
      } else {
        return(glue::glue("Argument {.arg {.strong {% argument %}}} is {.val NULL} and argument {.strong {.code ignore_null = FALSE}}", .open = "{%", .close = "%}"))
      }
    }
    if (!inherits(arguments[[argument]], target_classes) & !is.null(arguments[[argument]])) {
      return(glue::glue("Argument {.arg {.strong {% argument %}}} has class {.cls {% class(arguments[[argument]]) %}}, but must have {cli::qty(target_classes)} {?any of the} class{?es}: {.cls {target_classes}}", .open = "{%", .close = "%}"))
    }
  })
  names(failures) <- names(arguments)
  failures <- Filter(function(x) !is.null(x), failures)
  if (!all(sapply(failures, is.null))) {
    failures_out <- sapply(names(failures), FUN = function(argument) {
      if (!is.null(failures[[argument]])) {
        return(failures[[argument]])
      }
    })
    cli::cli_abort(c(
      "x" = "inheritance checks {cli::col_red('failed')} in function call: {.strong {.code {caller_func}}} due to following errors:",
      failures_out
    ))
  }
  cli::cli_alert_success("inheritance checks {col_green('passed')}")
}
```

File: R/make_inheritance_checks.R [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+make_inheritance_checks <- function(arguments, target_classes, ignore_null = T) {
+  caller_func <- deparse(sys.calls()[[sys.nframe() - 1]])
+  failures <- lapply(names(arguments), FUN = function(argument) {
+    if (is.null(arguments[[argument]])) {
+      if (ignore_null) {
+        cli_alert_info("{.fn make_inheritance_checks}: Argument {.arg { argument }} is {.val NULL}, but argument {.strong {.code ignore_null = TRUE}}.")
+        return(NULL)
+      } else {
+        return(glue("Argument {.arg {.strong {% argument %}}} is {.val NULL} and argument {.strong {.code ignore_null = FALSE}}", .open = "{%", .close = "%}"))
```
Comment (Moohan) [inline R/make_inheritance_checks.R:16]:
```suggestion
        return(glue::glue("{.arg {% argument %}} is {.val NULL} but must be {cli::qty(target_classes)} {?any of }{.cls {target_classes}}.", .open = "{%", .close = "%}"))
```

File: R/make_inheritance_checks.R [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+      }
+    }
+    if (!inherits(arguments[[argument]], target_classes) & !is.null(arguments[[argument]])) {
+      return(glue("Argument {.arg {.strong {% argument %}}} has class {.cls {% class(arguments[[argument]]) %}}, but must have {qty(target_classes)} {?any of the} class{?es}: {.cls {target_classes}}", .open = "{%", .close = "%}"))
```
Comment (Moohan) [inline R/make_inheritance_checks.R:20]:
```suggestion
      return(glue::glue("{.arg {% argument %}} has class {.cls {% class(arguments[[argument]]) %}}, but must be {cli::qty(target_classes)} {?any of }{.cls {target_classes}}.", .open = "{%", .close = "%}"))
```

File: R/make_inheritance_checks.R [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+  if (!all(sapply(failures, is.null))) {
+    failures_out <- sapply(names(failures), FUN = function(argument) {
+      if (!is.null(failures[[argument]])) {
+        return(failures[[argument]])
+      }
+    })
+    cli_abort(c(
+      "x" = "inheritance checks {col_red('failed')} in function call: {.strong {.code {caller_func}}} due to following errors:",
+      failures_out
+    ))
```
Comment (Moohan) [inline R/make_inheritance_checks.R:34]:
```suggestion
    cli::cli_abort(c(
      "x" = "{.strong {.code {caller_func}}} {cli::col_red('failed')}, check the arguments:",
      failures_out
    ))
```

**Review by Moohan — CHANGES_REQUESTED** [review]
Thanks for making those changes.

Your code didn't work for me as you have to specify the functions as `pkg::func` because `cli` (or other packages) aren't (necessarily) loaded when `phsmethods` is so the calls need to be explicit.

Can you make sure to test (`devtools::test()`) and check (`devtools::check()`) before committing please as these will highlight a lot of these errors?

Still to do:
- [ ] Make the function quieter. i.e. only print a message if there's a problem the user can fix.
- [ ] Make the tests pass - You will need to edit the test files `tests/testthat/test-age_calculate.R` and `tests/testthat/test-dob_from_chi.R`. These tests are currently testing for the 'old' error messages, but need to be amended to test for all cases of your new error messages.
- [ ] Make sure R CMD Check (`devtools::check()`) passes. - This will mean fixing the tests first but there may be some other errors, warnings or notes to be addressed.
- [ ] It would be great if you could add some specific tests for this new function.

**Comment by JoKPro** [comment]
I see, thanks for the feedback!

**Comment by JoKPro** [comment]
I implemented the tasks you mentioned and made the tests pass, hope this works!

**Review by Moohan — APPROVED** [review]
Looks good to me. All checks and tests pass (including new tests). 
Error messages seem to be what we want as well.

The only minor issue is that `glue` is now used but not imported i.e. it's not in the `DESCRIPTION` file. This can be fixed with `usethis::use_package("glue")` (could be done here or after merge to dev @Tina815). The other option would be to replace `glue::glue()` with `stringr::str_glue` as that's a wrapper anyway and we're already importing `stringr`.

**Pull request merged**